### ISSUE TITLE
Rule proposal: require SAFETY: comments for TSAsExpression

### ISSUE THREAD

**Issue by adrianbw** [typescript-eslint/typescript-eslint#8007]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

My rule would require a comment explaining why a TSAsExpression was safe. This is useful because people who later read the code will understand why the cast was made.

### Fail Cases

```TypeScript
const a = 'foo' as string;
foo(bar as string);
```

### Pass Cases

```TypeScript
// SAFETY: foo
const a = 'foo' as string;

foo(bar as string); // SAFETY: foo

/**
 * SAFETY: foo
 */
const a = 'foo' as string;
```

**Comment by bradzacher** [comment]
We already have this! If you turn on [consistent-type-assertions](https://typescript-eslint.io/rules/consistent-type-assertions) to `never` then it'll ban `as` assertions.

Then you must either not use an assertion or use a disable comment.
```ts
// eslint-disable-next-line @typescript-eslint/consistent-type-assertions -- SAFETY: foo
const a = 'foo' as string;

foo(bar as string); // eslint-disable-line @typescript-eslint/consistent-type-assertions -- SAFETY: foo

/**
 * SAFETY: foo
 */
const a = 'foo' as string;
```
[playground](https://typescript-eslint.io/play#ts=5.2.2&fileType=.tsx&code=PQKgBApgzgNglgOwC5gAJIJ4AdoGMBOcWSAtNPMsLgPYJRxRITImY4kCGUUE%2BSctKAC4wAbQBEvfNXziANGADeXHnwEIAyphgQRAcgQQAbrz0BfALpgQwALAAoBwBMIuGB3wQwNOigBm1NQiHAgYANzOru6e3oIoAEYewaERjmkOwMCQsIikTgwc8TokhgAepBRe6Nh4hMRkOZQ%2B9IzMpGwQnNy8-IJgJCRgGgCCAGIAogAqAJoiAdQOzSgcYAC8YHrzemBcYIyECADmqQ7zABSJ%2BDtQe0gHhwCUYWCZ2RR5BUWdlWgdUAREUjkXJUQQMJgsDpdVS9Oj9QYjCYzOaBBwZEAOYEsfJQQrFMoVRBVP4A%2BpYpCgujgtqsGrQnrqG4DByIqazMDzBw2RZxMDxNYbLbXW73VJAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

You can manually enforce disable comment descriptions at review time or you can use a rule like [`eslint-comments/require-description`](https://mysticatea.github.io/eslint-plugin-eslint-comments/rules/require-description.html) to enforce it

**Comment by wagenet** [comment]
@bradzacher I think this misunderstands the rule. The goal here is not to ban `as` assertions, the goal is to require people to explain the assertion. You suggest that we could do one-off disabling, but that doesn't address the use case either since 1) we do not want to encourage developers to do one-off disabling in our code and 2) with one off disabling it would still be easy to forget to add a SAFETY comment.

**Comment by bradzacher** [comment]
> with one off disabling it would still be easy to forget to add a SAFETY comment.

I addressed this in my comment - there's a lint rule for that.

> The goal here is not to ban as assertions, the goal is to require people to explain the assertion.

From a linting perspective - these are the same thing. A rule reports on an assertion and you either remove the assertion or you add a comment to silence the linter.

The only difference between the two is the content of the comment - where your suggestion is a bespoke form that is short, mine is a built-in form that is longer.

> we do not want to encourage developers to do one-off disabling 

I definitely understand that as someone who has been the steward of several massive codebases - but I also recognise that disables are just a part of writing code with a linter. Sometimes the linter is wrong or you need to intentionally violate the linter for a good reason.
 
Disable comments are just a tool to be used is all.

---

I definitely understand what you're asking for here - but the intent from my response is saying that  there already exists a solution that exists that will service the vast majority of codebases - so we won't be building this into our project.

**Comment by wagenet** [comment]
Thanks for elaborating. This is still something I'd like, but I definitely see where you're coming from.
